File: algorithms/496. Next Greater Element I/solution.py
class Solution(object):
    def nextGreaterElement(self, findNums, nums):
        """
        :type findNums: List[int]
        :type nums: List[int]
        :rtype: List[int]
        """
        # @yuxiangmusic
        desc_stack = []
        next_greater_map = {}
        for n in nums:
            while desc_stack and desc_stack[-1] < n:
                next_greater_map[desc_stack.pop()] = n
            desc_stack.append(n)
        return map(lambda x: next_greater_map.get(x, -1), findNums)

# 17 / 17 test cases passed.
# Status: Accepted
# Runtime: 52 ms
# 93.64%

File: algorithms/496. Next Greater Element I/solution.cc
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& findNums, vector<int>& nums) {
        stack<int> desc;
        unordered_map<int, int> nextGreater;
        for ( int n : nums ) {
            while ( desc.size() && desc.top() < n ) {
                nextGreater[desc.top()] = n;
                desc.pop();
            }
            desc.push(n);
        }
        vector<int> result;
        for ( int n : findNums ) {
            result.push_back(nextGreater.count(n) ? nextGreater[n] : -1);
        }
        return result;
    }
};
